pkg/che/che.go

```diff
@@ -23,7 +40,123 @@ type Client struct {
 func InitDefaultCheClient(cfg *crtcfg.Config, cl client.Client) {
 	DefaultClient = &Client{
 		config:     cfg,
+		httpClient: newHTTPClient(),
 		k8sClient:  cl,
 		tokenCache: newTokenCache(),
 	}
 }
+
+// UserExists returns true if the username exists, false if it doesn't and an error if there was problem with the request
+func (c *Client) UserExists(username string) (bool, error) {
+	reqData := url.Values{}
+	reqData.Set("name", username)
+	res, err := c.cheRequest(http.MethodGet, cheUserFindPath, reqData)
+	if err != nil {
+		return false, errors.Wrapf(err, "request to find Che user '%s' failed", username)
+	}
+	defer rest.CloseResponse(res)
+	if res.StatusCode == http.StatusOK {
+		return true, nil
+	} else if res.StatusCode == http.StatusNotFound {
+		return false, nil
+	}
+	return false, errors.Errorf("request to find Che user '%s' failed, Response status: '%s'", username, res.Status)
+}
+
+// GetUserIDByUsername returns the user ID that maps to the given username
+func (c *Client) GetUserIDByUsername(username string) (string, error) {
+	reqData := url.Values{}
+	reqData.Set("name", username)
+	res, err := c.cheRequest(http.MethodGet, cheUserFindPath, reqData)
+	if err != nil {
+		return "", errors.Wrapf(err, "unable to get Che user ID for user '%s'", username)
+	}
+	defer rest.CloseResponse(res)
+	if res.StatusCode != http.StatusOK {
+		resBody, _ := rest.ReadBody(res.Body)
+		err = errors.Errorf("unable to get Che user ID for user '%s', Response status: '%s' Body: '%s'", username, res.Status, resBody)
+		return "", err
+	}
+	cheUser, err := readCheUser(res)
+	if err != nil {
+		return "", errors.Wrapf(err, "unable to get Che user ID for user '%s'", username)
+	}
+	return cheUser.ID, err
+}
+
+// DeleteUser deletes the Che user with the given user ID
+func (c *Client) DeleteUser(userID string) error {
```

Comment:
Maybe we should make all these functions internal/not exposed and later add just one exposed function: DeleteUser(username) which would call all the internal functions. So the client (controller) just need to call that DeleteUser(username) and leave everything else to the che client.

Reply:
That's actually what I did initially, but I discovered it doesn't seem to be that simple because `UserExists` and `GetUserIDByUsername` can return 404 while the deletion is still in progress and since we need to get the UserID to do a retry of `DeleteUser` until it returns a 204 normally it was getting a bit messy to have everything handled by a `DeleteUser` function in che.go. This is the basic implementation I have (also what I used for the demo): https://github.com/rajivnathan/member-operator/blob/64cf4a336a8dcf416bf8ce6eee937b52b9a85c26/pkg/controller/useraccount/useraccount_controller.go#L653

It can probably be made a bit cleaner though, so I'm open to suggestions for improving that.

Reply:
Hm.. OK. Che keeps complicating our stuff :)
Let'd keep it as is then.